### PR TITLE
LPS-159826 | Add autom. tests for walkthrough

### PR DESCRIPTION
**Pull request by mateusralv**
Background:
Add StepCanCorrespondToCorrectStepId and CanRenderWhenStepIsBelowFold autom. tests

Test Plan:
WalkthroughSiteLevel#StepCanCorrespondToCorrectStepId
Given walkthrough enabled site-level
When step IDs are defined in the pages property
and when walkthrough steps correspond to the correct step ID on correct page
Then complete walkthrough process

WalkthroughSiteLevel#CanRenderWhenStepIsBelowFold
Given walkthrough enabled site-level
And Given a page width larger than the preview window
When enable popover mode
And When navigate to step that is below the fold
Then popover element is visible when rendered

JIRA Link:
https://issues.liferay.com/browse/LPS-159826
https://issues.liferay.com/browse/LPS-159844

### CODE REVIEW

**Comment by mateusralv** [comment]
ci:test:frontend

**Comment by mateusralv** [comment]
ci:test:frontend

**Comment by john-co** [comment]
Just started reviewing :)

**Comment by john-co** [comment]
ci:test:sf

File: modules/apps/frontend-js/frontend-js-test/src/testFunctional/tests/WalkthroughSiteLevel.testcase [hunk 1]
```diff
@@ -60,4 +64,87 @@ definition {
...
+				widgetName = "JS Walkthrough Sample");
+
+			JSONLayout.addPublicLayout(
+				groupName = "Site Name",
+				layoutName = "Walkthrough Page 2");
+
+			JSONLayout.addWidgetToPublicLayout(
+				groupName = "Site Name",
+				layoutName = "Walkthrough Page 2",
+				widgetName = "Documents and Media");
```
Comment (john-co) [inline modules/apps/frontend-js/frontend-js-test/src/testFunctional/tests/WalkthroughSiteLevel.testcase:93]: I'd try to reduce dependencies from other modules here... :thinking: 

Is there some more generic and decoupled portlet not tied to a product team?

File: modules/apps/frontend-js/frontend-js-test/src/testFunctional/tests/WalkthroughSiteLevel.testcase [hunk 2]
```diff
@@ -60,4 +64,87 @@ definition {
...
+
+			Walkthrough.enablePopoverMode();
+		}
+
+		task ("And When navigate to step that is below the fold") {
+			Walkthrough.goToNextStep(key_step = "2");
+		}
+
+		task ("Then popover element is visible when rendered") {
+			AssertElementPresent(
```
Comment (john-co) [inline modules/apps/frontend-js/frontend-js-test/src/testFunctional/tests/WalkthroughSiteLevel.testcase:114]: We're trying to assert that the element is visible, not present. AssertElementPresent can pass if the element is outside the viewport. I think you want AssertVisible, it will check if an element is within the viewport of browser window.

File: modules/apps/frontend-js/frontend-js-test/src/testFunctional/macros/Walkthrough.macro [hunk 3]
```diff
@@ -24,6 +24,19 @@ definition {
 		SystemSettings.saveConfiguration();
 	}
 
+	macro assertHighlightedSpecificObject {
```
Comment (john-co) [inline modules/apps/frontend-js/frontend-js-test/src/testFunctional/macros/Walkthrough.macro:27]: Cool

File: modules/apps/frontend-js/frontend-js-test/src/testFunctional/tests/WalkthroughSiteLevel.testcase (hunk 1 above)
Comment (mateusralv) [reply modules/apps/frontend-js/frontend-js-test/src/testFunctional/tests/WalkthroughSiteLevel.testcase:93]: Yep, I can try to add a image, seems good?

File: modules/apps/frontend-js/frontend-js-test/src/testFunctional/tests/WalkthroughSiteLevel.testcase [hunk 4]
```diff
@@ -60,4 +64,87 @@ definition {
...
+
+		task ("Given a page width larger than the preview window.") {
+			JSONLayout.addPublicLayout(
+				groupName = "Site Name",
+				layoutName = "Walkthrough Page 1");
+
+			JSONLayout.addWidgetToPublicLayout(
+				groupName = "Site Name",
+				layoutName = "Walkthrough Page 1",
+				widgetName = "JS Walkthrough Sample");
```
Comment (john-co) [inline modules/apps/frontend-js/frontend-js-test/src/testFunctional/tests/WalkthroughSiteLevel.testcase:84]: Looks like setupWalkthroughPage macro can be used here

File: modules/apps/frontend-js/frontend-js-test/src/testFunctional/tests/WalkthroughSiteLevel.testcase [hunk 5]
```diff
@@ -60,4 +64,87 @@ definition {
...
+
+			JSONLayout.addPublicLayout(
+				groupName = "Site Name",
+				layoutName = "Walkthrough Page 2");
+
```
Comment (john-co) [inline modules/apps/frontend-js/frontend-js-test/src/testFunctional/tests/WalkthroughSiteLevel.testcase:89]: Can probably refactor setupWalkthroughPage macro to include Page 2 also

File: portal-web/test/functional/com/liferay/portalweb/functions/AssertAttributeValue.function [hunk 6]
```diff
@@ -1,6 +1,18 @@
 @default = "assertAttributeValue"
 definition {
 
+	function assertAttributeNotPresent {
```
Comment (john-co) [inline portal-web/test/functional/com/liferay/portalweb/functions/AssertAttributeValue.function:4]: Nice

**Review by john-co — CHANGES_REQUESTED** [review]

File: modules/apps/frontend-js/frontend-js-test/src/testFunctional/tests/WalkthroughSiteLevel.testcase (hunk 1 above)
Comment (john-co) [reply modules/apps/frontend-js/frontend-js-test/src/testFunctional/tests/WalkthroughSiteLevel.testcase]: sounds good!

**Comment by mateusralv** [comment]
ci:test:sf

**Comment by mateusralv** [comment]
ci:test:frontend

File: modules/apps/frontend-js/frontend-js-test/src/testFunctional/tests/WalkthroughSiteLevel.testcase [hunk 7]
```diff
@@ -60,4 +66,69 @@ definition {
...
+		task ("Given walkthrough enabled site-level") {
+			Walkthrough.setupWalkthroughPage(
+				filename = "walkthrough_configuration_multiple_page.json",
+				siteName = "Site Name",
+				widgetList = "Blogs,JS Walkthrough Sample");
+		}
+
+		task ("Given a page width larger than the preview window.") {
+			SetWindowSize(value1 = "200,720");
```
Comment (mateusralv) [inline modules/apps/frontend-js/frontend-js-test/src/testFunctional/tests/WalkthroughSiteLevel.testcase:82]: We can not add an image with the addWidgetToPublicLayout macro decided to decrease the screen size and add a blog widget, seems good?

**Comment by mateusralv** [comment]
ci:test:frontend

**Comment by mateusralv** [comment]
The [failures](https://github.com/liferay-frontend/liferay-portal/pull/2764#issuecomment-1267377544) are not related. 🎉

File: modules/apps/frontend-js/frontend-js-test/src/testFunctional/macros/Walkthrough.macro [hunk 8]
```diff
@@ -78,26 +91,39 @@ definition {
...
+			if ("${filename}" == "walkthrough_configuration_multiple_page.json") {
+				JSONLayout.addPublicLayout(
+					groupName = "${siteName}",
+					layoutName = "Walkthrough Page 2");
+
+				for (var portletName : list "${widgetList}") {
+					JSONLayout.addWidgetToPublicLayout(
+						groupName = "${siteName}",
+						layoutName = "Walkthrough Page 2",
+						widgetName = "${portletName}");
```
Comment (john-co) [inline modules/apps/frontend-js/frontend-js-test/src/testFunctional/macros/Walkthrough.macro:118]: I don't like this for loop in setup walkthrough macro. We should just deploy walkthrough sample portlet. Adding an extra portlet is for 1 edge case and isn't worth including in this macro. General rule: If there's 3 or more instances of this then it can be considered to include for a new macro or refactoring a macro.

File: modules/apps/frontend-js/frontend-js-test/src/testFunctional/tests/WalkthroughSiteLevel.testcase [hunk 9]
```diff
@@ -60,4 +66,69 @@ definition {
...
+	@description = "LPS-159844. When a walkthrough step is defined and a step is below the fold, user can see popover is."
+	@priority = "3"
+	test CanRenderWhenStepIsBelowFold {
+		property portal.acceptance = "true";
+
+		task ("Given walkthrough enabled site-level") {
+			Walkthrough.setupWalkthroughPage(
+				filename = "walkthrough_configuration_multiple_page.json",
+				siteName = "Site Name",
+				widgetList = "Blogs,JS Walkthrough Sample");
```
Comment (john-co) [inline modules/apps/frontend-js/frontend-js-test/src/testFunctional/tests/WalkthroughSiteLevel.testcase:78]: Let's just leave this setupWalkthroughPage alone and not complicate it with other widgets.

You can just add another task prior to adding JS walkthrough sample ("Given vertically long ckeditor sample portlet") and use frontend ckeditor sample portlet. Maybe add Lorem ipsum text if need to extend the portlet down further.

File: modules/apps/frontend-js/frontend-js-test/src/testFunctional/tests/WalkthroughSiteLevel.testcase (hunk 7 above)
Comment (john-co) [reply modules/apps/frontend-js/frontend-js-test/src/testFunctional/tests/WalkthroughSiteLevel.testcase:82]: use frontend ckeditor sample portlet instead

**Review by john-co — CHANGES_REQUESTED** [review]

**Comment by mateusralv** [comment]
ci:test:frontend

**Review by john-co — APPROVED** [review]
Looks ok to forward but wanted to double check if you still wanted to refactor.

File: modules/apps/frontend-js/frontend-js-test/src/testFunctional/macros/Walkthrough.macro (hunk 8 above)
Comment (john-co) [reply modules/apps/frontend-js/frontend-js-test/src/testFunctional/macros/Walkthrough.macro]: Doesn't look like this change was made

File: modules/apps/frontend-js/frontend-js-test/src/testFunctional/tests/WalkthroughSiteLevel.testcase (hunk 9 above)
Comment (john-co) [reply modules/apps/frontend-js/frontend-js-test/src/testFunctional/tests/WalkthroughSiteLevel.testcase]: Doesn't look like this change was made

Comment (mateusralv) [reply modules/apps/frontend-js/frontend-js-test/src/testFunctional/tests/WalkthroughSiteLevel.testcase]: I kept the macro as it was, because I need the sample of the Walkthrough sample to be added after ckeditor sample to stay below and the scroll to be used. That is, add the page > add the ckeditor sample > add the Walkthrough sample.

Comment (mateusralv) [reply modules/apps/frontend-js/frontend-js-test/src/testFunctional/tests/WalkthroughSiteLevel.testcase]: The `JSONLayout.addWidgetToPublicLayout `macro adds the widget below if it is already on the page

**Comment by john-co** [comment]
ci:forward
